pkg/subscription/context.go

```diff
@@ -17,27 +18,51 @@ func NewInitialHttpRequestContext(r *http.Request) *InitialHttpRequestContext {
 	}
 }
 
-type subscriptionCancellations map[string]context.CancelFunc
+type subscriptionCancellations struct {
+	mu            sync.RWMutex
+	cancellations map[string]context.CancelFunc
+}
 
-func (sc subscriptionCancellations) AddWithParent(id string, parent context.Context) context.Context {
+func (sc *subscriptionCancellations) AddWithParent(id string, parent context.Context) context.Context {
 	ctx, cancelFunc := context.WithCancel(parent)
-	sc[id] = cancelFunc
+	sc.mu.Lock()
+	defer sc.mu.Unlock()
+	if sc.cancellations == nil {
+		sc.cancellations = make(map[string]context.CancelFunc)
+	}
+	sc.cancellations[id] = cancelFunc
 	return ctx
 }
 
-func (sc subscriptionCancellations) Cancel(id string) (ok bool) {
-	cancelFunc, ok := sc[id]
+func (sc *subscriptionCancellations) Cancel(id string) (ok bool) {
+	sc.mu.Lock()
+	defer sc.mu.Unlock()
+	cancelFunc, ok := sc.cancellations[id]
 	if !ok {
 		return false
 	}
 
 	cancelFunc()
-	delete(sc, id)
+	delete(sc.cancellations, id)
 	return true
 }
 
-func (sc subscriptionCancellations) CancelAll() {
-	for _, cancelFunc := range sc {
+func (sc *subscriptionCancellations) CancelAll() {
+	// Copy the values to temporary storage to avoid calling arbitrary functions
+	// with the lock held
+	sc.mu.RLock()
+	funcs := make([]context.CancelFunc, 0, len(sc.cancellations))
```

Comment:
Why is this needed?

Reply:
Oh, I see! At the moment I was rewriting that I didn't realize we have full control of the cancellation functions, so my reasoning was that some function might call into one of the other mutex guarded methods and cause a deadlock, but now I see that's not possible. I'll remove the copy.